### ISSUE TITLE
Redesign `AstBuilder`

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#23043]
## Proposed API

Currently:

```rs
let rest = self.ast.binding_rest_element(span, argument);
let boxed_rest = self.ast.alloc_binding_rest_element(span, argument);
let stmt = self.ast.statement_expression(span, expr);
```

Instead, define creation methods on AST types themselves:

```rs
let rest = BindingRestElement::new(span, argument, self);
let boxed_rest = BindingRestElement::boxed(span, argument, self);
let stmt = Statement::new_expression_statement(span, expr, self);
```

## Why?

1. Clearer syntax. No need to remember the names of hundreds of `AstBuilder` methods.
2. Less borrow-checker problems. Passing `self` as *last* param means `&mut self` can be used in earlier params without borrow checker complaining.
3. Support `NodeId`s.
4. Prevent `NodeId`s accidentally being omitted in transformer. `NodeId` creation is automatic, and impossible to forget.

## Why now?

1. Counting AST nodes in parser (#188) without hurting parser perf requires making `AstBuilder` not `Copy`. This will be a large change, producing huge code churn. It makes sense IMO to do the 2 changes in one go, and the proposed design should require less code re-organisation due to borrow-checker problems than removing `Copy` from `AstBuilder` alone.
2. Transformer needs to produce AST nodes with unique `NodeId`s (https://github.com/rolldown/rolldown/pull/9609).

## Implementation

* Create 2 different AST builder types for parser and transformer/minifier.
* Introduce `AstBuild` + `GetAstBuilder` traits.
* Implement `AstBuild` trait on AST builder types.
* Implement `GetAstBuilder` on e.g. `ParserImpl`, `TraverseCtx`.
* Codegen methods on AST types which are generic over any `GetAstBuilder`.

```rs
// --- Ast builders ---

trait AstBuild<'a>: GetAllocator<'a> {
    fn node_id(&self) -> NodeId;
}

trait GetAstBuilder<'a> {
    /// The [`AstBuild`]er type that this provides access to.
    type Builder: AstBuild<'a>;

    /// Get the [`AstBuild`]er.
    fn builder(&self) -> &Self::Builder;
}

// For parser
pub struct AstBuilder<'a> {
    allocator: &'a Allocator,
    nodes_count: Cell<u32>,
}

impl<'a> AstBuild<'a> for AstBuilder<'a> {
    fn node_id(&self) -> NodeId {
        self.nodes_count.set(nodes_count.get() + 1);
        NodeId::DUMMY
    }
}

impl<'a> GetAstBuilder<'a> for AstBuilder<'a> {
    type Builder = Self;
    fn builder(&self) -> &Self {
        self
    }
}

impl<'a> GetAllocator<'a> for AstBuilder<'a> {
    fn allocator(&self) -> &'a Allocator {
        self.allocator
    }
}

// For transformer, minifier
struct AstBuilderWithNodeIds<'a> {
    allocator: &'a Allocator,
    next_node_id: Cell<u32>,
}

impl<'a> AstBuild<'a> for AstBuilderWithNodeIds<'a> {
    fn node_id(&self) -> NodeId {
        let node_id = self.next_node_id.get();
        self.next_node_id.set(node_id + 1);
        NodeId::from_raw(node_id)
    }
}

impl<'a> GetAstBuilder<'a> for AstBuilderWithNodeIds<'a> {
    type Builder = Self;
    fn builder(&self) -> &Self {
        self
    }
}

impl<'a> GetAllocator<'a> for AstBuilderWithNodeIds<'a> {
    fn allocator(&self) -> &'a Allocator {
        self.allocator
    }
}

// --- Accessors ---

// For parser
impl<'a> GetAstBuilder<'a> for ParserImpl<'a> {
    type Builder = AstBuilder<'a>;
    fn builder(&self) -> &AstBuilder<'a> {
        &self.ast
    }
}

impl<'a> GetAllocator<'a> for ParserImpl<'a> {
    fn allocator(&self) -> &'a Allocator {
        self.ast.allocator()
    }
}

// For transformer
impl<'a> GetAstBuilder<'a> for TraverseCtx<'a> {
    type Builder = AstBuilderWithNodeIds<'a>;
    fn builder(&self) -> &AstBuilderWithNodeIds<'a> {
        &self.ast
    }
}

impl<'a> GetAllocator<'a> for TraverseCtx<'a> {
    fn allocator(&self) -> &'a Allocator {
        self.ast.allocator()
    }
}

// --- AST type builder methods ---

impl<'a> BindingRestElement<'a> {
    pub fn new<B>(span: Span, argument: BindingPattern<'a>, builder: &B) -> Self
        where B: GetAstBuilder<'a>,
    {
        let builder = builder.builder();
        BindingRestElement { node_id: builder.node_id(), span, argument }
    }

    pub fn boxed<B>(span: Span, argument: BindingPattern<'a>, builder: &B) -> Box<'a, Self>
        where B: GetAstBuilder<'a>,
    {
        let builder = builder.builder();
        Box::new_in(Self::new(span, argument, builder), builder.allocator())
    }
}

impl<'a> Statement<'a> {
    pub fn expression_statement<B>(span: Span, expr: Expression<'a>, builder: &B) -> Self
        where B: GetAstBuilder<'a>,
    {
        Statement::ExpressionStatement(ExpressionStatement::boxed(span, expr, builder))
    }
}
```

### Separation of concerns

This design separates creation of `NodeId`s / counting nodes from the builder methods. No need to generate 2 separate `AstBuilder`s with 100s of methods each for the 2 use cases.

### Flexibility

The `GetAstBuilder` implementations allow flexible calling patterns:

```rs
// Short version
let rest = BindingRestElement::new(span, argument, self);
// Longer version if another field of `self` is mutably borrowed
let rest = BindingRestElement::new(span, argument, &self.ast);
```

### Future-proof

This API clears the way for moving to taking a `&mut` ref in future, to get rid of the `Cell` in `AstBuilder` (more performant).

## Migration strategy

* Put current `AstBuilder` behind `ast_builder` Cargo feature (enabled by default).
* Generate new methods with `ast_tools` codegen.
* Put them behind `ast_builder_new` Cargo feature.
* LLM create ast-grep transform to convert old to new syntax.
* Incrementally convert crates from old to new. Once a crate is converted, disable `ast_builder` feature for that crate, to verify conversion is complete.
* Once all Oxc crates and downstream consumers are migrated to new APIs, remove old `AstBuilder`.

Despite a lot of code churn, I would hope that LLM + ast-grep would make it pretty swift and painless.

**Comment by camchenry** [comment]
I like the `let rest = BindingRestElement::new(span, argument, self)`, that seems much nicer than what we currently have. I wonder if there is a way to enforce (maybe via ast-grep) that there is no direct building of AST nodes like `BindingRestElement { ... }`, without using `new()`. Seems like it would be preferable to ensure all code paths go through the new methods.

In general, I like the direction of this, giving a `trait` for `AstGenerator` provides us more flexibility with inserting code into the AST creation process which gives us more options in the future. One concern I have though with the parser version of implementation is that we don't just want to count total nodes, but also count the maximum number of potential scopes that might be created, as well as symbols and references. Those counts will vary with the type of node that we're creating, which I don't see a way to count that currently.

Could we add another function to the trait that allows getting a node ID and requires passing a node kind? or modify the existing function to require passing a `AstType`? There maybe other options too, those are just two that came to mind.

**Option 1**: Add separate method

```rust
trait AstGenerator<'a> {
    fn allocator(&self) -> &'a Allocator;
    fn node_id(&self) -> NodeId;
    fn typed_node_id(&self, ast_type: AstType) -> NodeId;
}

// For parser
pub struct AstBuilder<'a> {
    allocator: &'a Allocator,
    nodes_count: Cell<u32>,
    scopes_count: Cell<u32>,
}

impl<'a> AstGenerator<'a> for AstBuilder<'a> {
    fn allocator(&self) -> &'a Allocator {
        self.allocator
    }

    fn node_id(&self) -> NodeId {
        self.nodes_count.set(self.nodes_count.get() + 1);
        NodeId::DUMMY
    }

    fn typed_node_id(&self, ast_type: AstType) -> NodeId {
        if ast_type.may_create_scope() {
            self.scopes_count.set(self.scopes_count.get() + 1);
        }
        self.node_id()
    }
}
```

**Option 2**: Add `AstType` to existing method:


```rust
trait AstGenerator<'a> {
    fn allocator(&self) -> &'a Allocator;
    fn node_id(&self, ast_type: AstType) -> NodeId;
}

// For parser
pub struct AstBuilder<'a> {
    allocator: &'a Allocator,
    nodes_count: Cell<u32>,
    scopes_count: Cell<u32>,
}

impl<'a> AstGenerator<'a> for AstBuilder<'a> {
    fn allocator(&self) -> &'a Allocator {
        self.allocator
    }

    fn node_id(&self, ast_type: AstType) -> NodeId {
        if ast_type.may_create_scope() {
            self.scopes_count.set(self.scopes_count.get() + 1);
        }
        self.nodes_count.set(self.nodes_count.get() + 1);
        NodeId::DUMMY
    }
}
```

This has the downside of forcing us to be explicit when getting a node_id:

```rust
impl<'a> BindingRestElement<'a> {
    pub fn new<A, G>(span, argument, accessor: &A) -> Self
        where
            A: AsRef<G>,
            G: AstGenerator<'a>,
    {
        let builder = accessor.as_ref();
        BindingRestElement { node_id: builder.node_id(AstType::BindingRestElement), span, argument }
        //                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    }
}
```

It doesn't seem _too_ bad though, and should be pretty easily to generate the code for that with AI or whatever.

**Comment by overlookmotel** [comment]
Thanks for the thoughtful feedback, and thanks for coming back so swiftly.

> I wonder if there is a way to enforce (maybe via ast-grep) that there is no direct building of AST nodes

We should definitely do that.

One way to enforce it at compile time is to insert a non-`pub` ZST field into every struct like `pub(crate) _dummy: ()` (we could insert it in `#[ast]` macro). Then only code within `oxc_ast` crate can construct AST nodes directly. We'd have to audit all code in `oxc_ast` crate to make sure it always goes via builder methods, but code in `oxc_transformer` etc would be unable to.

> we don't just want to count total nodes, but also count the maximum number of potential scopes...

Ah yes. I omitted that for brevity.

What I imagined was scope IDs working the same as node IDs. My thought was `AstGenerator` trait would also have a `scope_id` method, and `BlockStatement::new` etc would call `builder.scope_id()` to populate `scope_id: ScopeId` fields in structs which have that field (and increment counter in parser `AstBuilder`). Ditto `SymbolId` and `ReferenceId`.

If we want to force transformer code to provide `ScopeId` etc itself, we could add additional methods e.g.:

```rs
impl<'a> Program<'a> {
    pub fn new<A, G>(body: Vec<'a, Statement<'a>>, accessor: &A) -> Self
        where
            A: AsRef<G>,
            G: AstGeneratorWithoutSemanticIds<'a>, // Without
    {
        // ...
    }

    pub fn new_with_scope_id<A, G>(body: Vec<'a, Statement<'a>>, scope_id: ScopeId, accessor: &A) -> Self
        where
            A: AsRef<G>,
            G: AstGeneratorWithSemanticIds<'a>, // With
    {
        // ...
    }
}

trait AstGeneratorWithoutSemanticIds<'a>: AstGenerator<'a> {}
trait AstGeneratorWithSemanticIds<'a>: AstGenerator<'a> {}
```

* `AstGeneratorWithoutSemanticIds` would be implemented for `AstBuilder` (parser).
* `AstGeneratorWithSemanticIds` would be implemented for `AstBuilderWithNodeIds` (transformer).
* `new` would call `AstGenerator::scope_id` which would increment the scope counter.
* `new_with_scope_id` wouldn't call `AstGenerator::scope_id`, and use the provided `ScopeId`.

I think *probably* it's simpler keeping `AstType` out of it (and not relying on inlining to const-fold `ast_type.may_create_scope()`), but that's arguable.

What I am keen on is preventing accidental incorrect use at compile time - make it impossible to forget to create a `NodeId` / `ScopeId` / etc where one is required.

Either of your solutions, or mine I think would solve the problem. All the AST type methods (`BindingRestElement::new` etc) will be codegenned, so it's not difficult to do it however we want, and make sure it's always accurate.

**Comment by overlookmotel** [comment]
Probably simpler option for forcing all AST node construction to go via builder methods is to mark all AST structs `#[non_exhaustive]`. That has the same effect as the non-pub dummy field I mentioned above. Again, we could add it automatically in `#[ast]` macro.

This could be done as a separate change, before the main change discussed in this issue.

**Comment by Boshen** [comment]
I approve.

**Comment by camchenry** [comment]
> Probably simpler option for forcing all AST node construction to go via builder methods is to mark all AST structs `#[non_exhaustive]`. That has the same effect as the non-pub dummy field I mentioned above. Again, we could add it automatically in `#[ast]` macro.

Oh I like that solution, that seems like a good approach to me.



> Ah yes. I omitted that for brevity.
> 
> What I imagined was scope IDs working the same as node IDs. My thought was `AstGenerator` trait would also have a `scope_id` method, and `BlockStatement::new` etc would call `builder.scope_id()` to populate `scope_id: ScopeId` fields in structs which have that field (and increment counter in parser `AstBuilder`). Ditto `SymbolId` and `ReferenceId`.

> What I am keen on is preventing accidental incorrect use at compile time - make it impossible to forget to create a `NodeId` / `ScopeId` / etc where one is required.
> 
> Either of your solutions, or mine I think would solve the problem. All the AST type methods (`BindingRestElement::new` etc) will be codegenned, so it's not difficult to do it however we want, and make sure it's always accurate.

Fair enough, and I don't think this design precludes us from implementing this in the future. I say let's move forward with this and we can add the AST node counting stuff in the future.

**Comment by hyfdev** [comment]
From the rolldown side, on whether further churn to the construction API is a concern — it isn't, and here's why:

- We're consolidating **all** of rolldown's oxc AST construction through a single point: oxc's `AstBuilder`, plus a thin `AstBuilderExt` extension trait for rolldown-specific node patterns.
- That single point becomes an insulation boundary. If `AstBuilder` is later reshaped or **removed**, rolldown's call sites are unaffected.
- In that case rolldown would just re-host its own `AstBuilder` (same construction surface) at that one point to serve its own node-construction needs.
- We're migrating rolldown's internal call sites onto these methods now, so any future upstream change stays contained to one layer instead of rippling across hundreds of sites.

So +1 to this direction — more changes down the line are fine on our end.

Design doc / plan: https://github.com/rolldown/rolldown/pull/9673

**Comment by overlookmotel** [comment]
#23876 Completes the new `AstBuilder` interface. All Oxc's crates are migrated over.

There is a Cargo feature `disable_old_builder` (enabled in tests in all Oxc's crates which use `AstBuilder`). This feature removes the old legacy methods from `AstBuilder`, and also removes `#[derive(Copy, Clone)]` from `AstBuilder` type.

Downstream consumers should migrate over to the new `AstBuilder` interface as soon as possible, and then enable the `disable_old_builder` feature to prevent new code being introduced to their project which relies on the old interface.

We will remove the old methods in around 2 weeks, once major downstream consumers have migrated.

Migration of Oxc's own crates was performed primarily with a codemod which is available on [overlookmotel/ast-builder-codemod](https://github.com/oxc-project/oxc/tree/overlookmotel/ast-builder-codemod) branch. It will likely need some modifications to work for other projects, but coding agents should be able to figure it out quite easily.

The immediate breaking change (which isn't behind `disable_old_builder` feature) is that methods of `Box` and `Vec` now take a different type.

```rs
use oxc_allocator::{Allocator, ArenaVec, ArenaBox};

let allocator = Allocator::new();

// Before
let boxed = ArenaBox::new_in(&allocator);
let vec = ArenaVec::new_in(&allocator);

// After - methods take double-reference
let allocator = &allocator;
let boxed = ArenaBox::new_in(&allocator);
let vec = ArenaVec::new_in(&allocator);
```

Some more details in the doc comment here: https://github.com/oxc-project/oxc/blob/0ab64ecfbdfca1ee7f9f9f6411e18c535876598b/crates/oxc_ast/src/builder/mod.rs#L1-L96
